File: packages/sdk-py/src/enclii_sdk/resources/jobs.py
"""Timetable jobs: cron jobs and one-off jobs.

Note: the switchyard-api cron-job endpoints currently return 501 (stub).
The SDK is shipped anyway so consumers can start wiring call sites —
the shape will match when the handlers are implemented.
"""

from __future__ import annotations

from enclii_sdk.models.jobs import (
    CreateCronJobRequest,
    CreateOneOffJobRequest,
    CronJob,
    CronJobRun,
)
from enclii_sdk.resources._base import Resource


class JobsResource(Resource):
    """Cron jobs and one-off jobs (``timetable`` tag)."""

    async def create_cron(
        self,
        project_slug: str,
        *,
        name: str,
        schedule: str,
        command: list[str],
        image: str,
        environment_name: str = "production",
    ) -> CronJob:
        """Create a cron job.

        API: ``POST /v1/projects/{slug}/cron-jobs``. Requires ``developer``.
        """
        request = CreateCronJobRequest(
            name=name,
            schedule=schedule,
            command=command,
            image=image,
            environment_name=environment_name,
        )
        data = await self._client.post(
            f"/v1/projects/{project_slug}/cron-jobs",
            json_body=request.model_dump(mode="json"),
        )
        return CronJob.model_validate(data)

    async def list_cron(self, project_slug: str) -> list[CronJob]:
        """List cron jobs for a project."""
        data = await self._client.get(f"/v1/projects/{project_slug}/cron-jobs")
        items = data if isinstance(data, list) else data.get("cron_jobs", [])
        return [CronJob.model_validate(j) for j in items]

    async def get_cron(self, job_id: str) -> CronJob:
        """Fetch a cron job by id."""
        data = await self._client.get(f"/v1/cron-jobs/{job_id}")
        return CronJob.model_validate(data)

    async def delete_cron(self, job_id: str) -> None:
        """Delete a cron job. Requires ``admin``."""
        await self._client.delete(f"/v1/cron-jobs/{job_id}")

    async def list_cron_runs(self, job_id: str) -> list[CronJobRun]:
        """List historical runs of a cron job."""
        data = await self._client.get(f"/v1/cron-jobs/{job_id}/runs")
        items = data if isinstance(data, list) else data.get("runs", [])
        return [CronJobRun.model_validate(r) for r in items]

    async def create_one_off(
        self,
        project_slug: str,
        *,
        name: str,
        command: list[str],
        image: str,
        environment_name: str = "production",
    ) -> CronJob:
        """Create a one-off job (runs once and exits).

        API: ``POST /v1/projects/{slug}/one-off-jobs``. Requires ``developer``.
        """
        request = CreateOneOffJobRequest(
            name=name,
            command=command,
            image=image,
            environment_name=environment_name,
        )
        data = await self._client.post(
            f"/v1/projects/{project_slug}/one-off-jobs",
            json_body=request.model_dump(mode="json"),
        )
        return CronJob.model_validate(data)

File: packages/sdk-ts/src/resources/jobs.ts
import type { EncliiClient } from '../client';
import type { Page, UnpagedListOptions } from '../types';
import type {
  CreateCronJobRequest,
  CreateOneOffJobRequest,
  CronJob,
  CronJobRun,
  OneOffJob,
} from '../types-ops';

/**
 * Timetable — cron and one-off scheduled jobs
 * (`apps/switchyard-api/internal/api/timetable_handlers.go`).
 *
 * No list endpoint here pages: `listCron()` returns every cron job, while
 * `listCronRuns()` and `listOneOff()` return only the 50 most recent rows.
 */
export class JobsResource {
  constructor(private readonly client: EncliiClient) {}

  // ---------------------------------------------------------------------------
  // Cron
  // ---------------------------------------------------------------------------

  async listCron(
    projectSlug: string,
    _options: UnpagedListOptions = {},
  ): Promise<Page<CronJob>> {
    const resp = await this.client.get<{ cron_jobs: CronJob[]; total: number }>(
      `/projects/${encodeURIComponent(projectSlug)}/cron-jobs`,
    );
    return { data: resp.cron_jobs ?? [], nextCursor: null };
  }

  async getCron(jobId: string): Promise<CronJob> {
    return this.client.get<CronJob>(
      `/cron-jobs/${encodeURIComponent(jobId)}`,
    );
  }

  /** Create a cron job. The API wraps it as `{ cron_job, message }`. */
  async createCron(
    projectSlug: string,
    input: CreateCronJobRequest,
  ): Promise<CronJob> {
    const resp = await this.client.post<{ cron_job: CronJob; message: string }>(
      `/projects/${encodeURIComponent(projectSlug)}/cron-jobs`,
      input,
    );
    return resp.cron_job;
  }

  async updateCron(
    jobId: string,
    input: Partial<CreateCronJobRequest> & {
      suspended?: boolean;
    },
  ): Promise<CronJob> {
    return this.client.patch<CronJob>(
      `/cron-jobs/${encodeURIComponent(jobId)}`,
      input,
    );
  }

  async deleteCron(jobId: string): Promise<void> {
    await this.client.del(`/cron-jobs/${encodeURIComponent(jobId)}`);
  }

  /** The 50 most recent runs of a cron job; older runs are not reachable. */
  async listCronRuns(
    jobId: string,
    _options: UnpagedListOptions = {},
  ): Promise<Page<CronJobRun>> {
    const resp = await this.client.get<{ runs: CronJobRun[]; total: number }>(
      `/cron-jobs/${encodeURIComponent(jobId)}/runs`,
    );
    return { data: resp.runs ?? [], nextCursor: null };
  }

  // ---------------------------------------------------------------------------
  // One-off
  // ---------------------------------------------------------------------------

  /** The 50 most recent one-off jobs of a project; older jobs are not reachable. */
  async listOneOff(
    projectSlug: string,
    _options: UnpagedListOptions = {},
  ): Promise<Page<OneOffJob>> {
    const resp = await this.client.get<{
      one_off_jobs: OneOffJob[];
      total: number;
    }>(`/projects/${encodeURIComponent(projectSlug)}/one-off-jobs`);
    return { data: resp.one_off_jobs ?? [], nextCursor: null };
  }

  /** Create a one-off job. The API wraps it as `{ one_off_job, message }`. */
  async createOneOff(
    projectSlug: string,
    input: CreateOneOffJobRequest,
  ): Promise<OneOffJob> {
    const resp = await this.client.post<{
      one_off_job: OneOffJob;
      message: string;
    }>(`/projects/${encodeURIComponent(projectSlug)}/one-off-jobs`, input);
    return resp.one_off_job;
  }
}
